zone/tests/zlogin_test.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 Oxide Computer Company
+
+use std::path::Path;
+use zone::{Adm, Config, CreationOptions, Zlogin};
+
+const PFEXEC: &str = "/bin/pfexec";
+
+// This test is marked with ignore because it involves booting a newly created
+// zone which can take over a minute. This test assumes that the sparse brand is
+// installed, if not use `pkg install brand/sparse`.
+#[test]
+#[ignore]
+fn test_zlogin() {
+    // Setup zfs pool for zone.
+    zfs_zonetest_create();
+
+    let name = "zexec";
+    let path = Path::new("/zonetest/zexec");
+
+    // Create a zone.
+    let mut cfg = Config::create(name, true, CreationOptions::Default);
+    cfg.get_global()
+        .set_path(path)
+        .set_autoboot(true)
+        .set_brand("sparse");
+    cfg.run().unwrap();
+
+    // Install and boot zone, cleaning up on failure.
+    let mut adm = Adm::new(name);
+    adm.install(&[]).unwrap_or_else(|e| {
+        cfg.delete(true).run().unwrap();
+        zfs_zonetest_destroy();
+        panic!("{}", e);
```

Comment:
Looking at all these unwinds, I think some of them are potentially incorrect.

- cfg.delete should *always* run after Config::create
- zfs_zonetest_destroy should *always* run after zfs_zonetest_create (I don't think this happens if "Config::create" fails)
- adm.uninstall should *always* run after adm.install
- adm.halt should *always* run after adm.boot

I went ahead and made a "defer" struct for you here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=239c8ac8377a40d63d338afce028499a

If you want, this might simplify some of the unwinding.

As an example:

```rust

    zfs_zonetest_create();
    let _unwind_zonetest_create = Defer::new(|| zfs_zonetest_destroy());
    ...
    let mut cfg = Config::create(name, true, CreationOptions::Default);
    ...
    cfg.run().unwrap();
    let _unwind_config_create = Defer::new(|| cfg.delete(true).run().unwrap());
```

Reply:
Awesome. I was thinking while writing this how `defer` from Go would be really nice here, but my Rust-fu is not quite there yet. Update coming shortly.